File: two pointers/max_num_of_ksum_pairs.java
import java.util.Arrays;
// two pointers and sorting
// tc O(nlogn), sc O(1)
public class max_num_of_ksum_pairs {
    public int maxOperations(int[] nums, int k) {
        Arrays.sort(nums);
        int l=0, r=nums.length-1, total_count=0;
        while(l<r){
            int curSum = nums[l]+nums[r];
            if(curSum<k){
                l++;
            }
            else if(curSum>k){
                r--;
            }
            else{
                total_count++;
                l++;
                r--;
            }
        }
        return total_count;
    }
}

File: two pointers/max_num_of_ksum_pairs.py
from typing import List


# two pointers and sorting
# tc O(nlogn), sc O(1)
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        nums.sort()
        l, r = 0, len(nums) - 1
        total_count = 0
        while l < r:
            _sum = nums[l] + nums[r]
            if _sum < k:
                l += 1
            elif _sum > k:
                r -= 1
            else:
                total_count += 1
                l += 1
                r -= 1
        return total_count
